tests/unit/test_charm.py

```diff
@@ -16,61 +16,63 @@
 
 
 def read_file(path: str) -> str:
-    """Read a file and returns as a string.
-
-    Args:
-        path (str): path to the file.
-
-    Returns:
-        str: content of the file.
-    """
+    """Read a file and returns as a string."""
     with open(path, "r") as f:
         content = f.read()
     return content
 
 
-class TestCharm(unittest.TestCase):
-    @patch("charm.UPFNetwork")
-    @patch("charm.Machine")
-    def setUp(self, patch_machine, patch_network):
+class TestCharm:
+    patcher_upf_network = patch("charm.UPFNetwork")
+    patcher_machine = patch("charm.Machine")
+    patcher_snap_cache = patch("charm.SnapCache")
+    patcher_pfcp_port = patch("charm.PFCP_PORT", TEST_PFCP_PORT)
+
+    @pytest.fixture()
+    def setup(self):
         self.mock_machine = MagicMock()
         self.mock_machine.pull.return_value = ""
         self.mock_process = MagicMock()
         self.mock_process.wait_output.return_value = ("Flags: avx2 rdrand", "")
         self.mock_machine.exec.return_value = self.mock_process
-        patch_machine.return_value = self.mock_machine
+        mock_machine = TestCharm.patcher_machine.start()
+        mock_machine.return_value = self.mock_machine
         self.mock_upf_network = MagicMock()
         self.mock_upf_network.get_invalid_network_interfaces.return_value = []
         self.mock_upf_network.core_interface.get_ip_address.return_value = "192.168.250.3"
-        patch_network.return_value = self.mock_upf_network
+        mock_upf_network = TestCharm.patcher_upf_network.start()
+        mock_upf_network.return_value = self.mock_upf_network
+        self.mock_snap_cache = TestCharm.patcher_snap_cache.start()
+        self.mock_pfcp_port = TestCharm.patcher_pfcp_port.start()
+
+    @staticmethod
+    def teardown() -> None:
+        patch.stopall()
+
+    @pytest.fixture(autouse=True)
+    def create_harness(self, setup, request):
         self.harness = ops.testing.Harness(SdcoreUpfCharm)
-        self.addCleanup(self.harness.cleanup)
+        self.harness.set_model_name(name="whatever")
+        self.harness.set_leader(is_leader=True)
         self.harness.begin()
+        yield self.harness
+        self.harness.cleanup()
+        request.addfinalizer(self.teardown)
 
     def test_given_unit_is_not_leader_when_evaluate_status_then_status_is_blocked(self):
         self.harness.set_leader(False)
 
         self.harness.evaluate_status()
 
-        self.assertEqual(
-            self.harness.model.unit.status,
-            ops.BlockedStatus("Scaling is not implemented for this charm"),
-        )
+        assert self.harness.model.unit.status == ops.BlockedStatus("Scaling is not implemented for this charm")  # noqa: E501
 
-    @patch("charm.UPFNetwork")
-    @patch("charm.SnapCache")
     def test_given_upf_snap_uninstalled_when_configure_then_upf_snap_installed(
-        self, mock_snap_cache, patch_network
+        self
     ):
         self.harness.set_leader(is_leader=True)
```

Comment:
Do we need to set the leader in every test? We already do it while creating the harness.